server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -43,8 +52,44 @@ public DesiredBalanceShardsAllocator(
             protected void processInput(DesiredBalanceInput desiredBalanceInput) {
                 if (desiredBalanceService.updateDesiredBalanceAndReroute(desiredBalanceInput, this::isFresh)) {
                     pendingReroute = true;
-                    rerouteServiceSupplier.get()
-                        .reroute("desired balance changed", Priority.NORMAL, ActionListener.wrap(() -> pendingReroute = false));
+                    boolean isFreshInput = isFresh(desiredBalanceInput);
+                    var listener = new ActionListener<ClusterState>() {
+                        @Override
+                        public void onResponse(ClusterState clusterState) {
+                            if (isFreshInput) {
+                                pendingReroute = false;
```

Comment:
We should clear this flag whether we're triggering listeners or not I think?

Reply:
Oh wait I see, the whole point is that we don't want to be considered idle in this case. The `pendingReroute` flag will go away soon I guess, it's only used in a few tests to implement an artificial wait that shouldn't be necessary.

Reply:
Correct. I am also going to remove it in a following pull requests where I will be fixing remaining ITs that are still depending on it.